Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift

```diff
@@ -0,0 +1,284 @@
+    }
+
+    func makeValueHistogram(name: String, unit: String? = nil, description: String? = nil, labels: [(String, String)], buckets: [Double]) -> ValueHistogram {
+        makeValueHistogram(name: name, unit: unit, description: description, attributes: Set(labels), buckets: buckets)
+    }
+}
+
+struct InstrumentIdentifier: Equatable, Hashable, Sendable {
+    var name: String
+    var unit: String?
+    var description: String?
+    enum InstrumentKind { case counter, gauge, histogram }
+    var kind: InstrumentKind
+
+    private init(name: String, unit: String? = nil, description: String? = nil, kind: InstrumentKind) {
+        self.name = name.lowercased()
+        self.unit = unit
+        self.description = description
+        self.kind = kind
+    }
+
+    static func counter(name: String, unit: String? = nil, description: String? = nil) -> Self {
+        self.init(name: name, unit: unit, description: description, kind: .counter)
+    }
+
+    static func gauge(name: String, unit: String? = nil, description: String? = nil) -> Self {
+        self.init(name: name, unit: unit, description: description, kind: .gauge)
+    }
+
+    static func histogram(name: String, unit: String? = nil, description: String? = nil) -> Self {
+        self.init(name: name, unit: unit, description: description, kind: .histogram)
+    }
+}
+
+protocol IdentifiableInstrument {
+    var instrumentIdentifier: InstrumentIdentifier { get }
+}
+
+extension Counter: IdentifiableInstrument {
+    var instrumentIdentifier: InstrumentIdentifier { .counter(name: name, unit: unit, description: description) }
+}
+
+extension Gauge: IdentifiableInstrument {
+    var instrumentIdentifier: InstrumentIdentifier { .gauge(name: name, unit: unit, description: description) }
+}
+
+extension Histogram: IdentifiableInstrument {
+    var instrumentIdentifier: InstrumentIdentifier { .histogram(name: name, unit: unit, description: description) }
+}
+
+protocol DuplicateRegistrationHandler: Sendable {
+    func handle(newRegistration: InstrumentIdentifier, existingRegistrations: Set<InstrumentIdentifier>)
+}
+
+struct FatalErrorDuplicateRegistrationHandler: DuplicateRegistrationHandler {
+    func handle(newRegistration: InstrumentIdentifier, existingRegistrations: Set<InstrumentIdentifier>) {
+        fatalError("""
+        Duplicate instrument registration for name: \(newRegistration.name)
+        ---
+        Instrument \(newRegistration) conflicts with existing instruments: \(existingRegistrations).
+        """)
+    }
+}
+
+struct WarningDuplicateRegistrationHandler: DuplicateRegistrationHandler {
+    let logger: Logger
+
+    func handle(newRegistration: InstrumentIdentifier, existingRegistrations: Set<InstrumentIdentifier>) {
+        self.logger.warning("Duplicate instrument registration", metadata: [
+            "newRegistration": "\(newRegistration)",
+            "existingRegistrations": "\(existingRegistrations)",
```

Comment:
```suggestion
            "existingRegistrations": .array(existingRegistrations.map { "\($0)" }),
```

Nitpicky, but I think this might read a bit nicer in some log handlers.

Reply:
Sure, done.